lib/launcher.rb

```diff
@@ -95,6 +95,7 @@ def https_app(https_port, plugins = https_plugins)
         :SSLVerifyClient => OpenSSL::SSL::VERIFY_PEER,
         :SSLPrivateKey => load_ssl_private_key(settings.ssl_private_key),
         :SSLCertificate => load_ssl_certificate(settings.ssl_certificate),
+        :SSLExtraChainCert => File.open(settings.foreman_ssl_ca).read.each_line.slice_after(/END CERTIFICATE/).map(&:join).map { |pem| OpenSSL::X509::Certificate.new(pem) },
```

Comment:
Looking at https://github.com/ruby/openssl/pull/441 then this should work:
```suggestion
        :SSLExtraChainCert => OpenSSL::X509::Certificate.load_file(fullchain_path),
```

But I didn't check versions etc.

Reply:
I found that too, but the method is not present on the ruby version present on rhel8 (ruby 2.7). That's why I went for this approach.

Checking the documentation, first occurrence of the methods load and load_file is on Ruby 3.1

Reply:
It is indeed listed in version 3.0.0 of the openssl library: https://github.com/ruby/openssl/blob/master/History.md#version-300

Perhaps worth a TODO?

Reply:
I thought more about it and I'd prefer some helper method (like `read_certificate_file`) that does the heavy lifting and effectively mimics the `load_file` method. Having such a long line in here makes it hard to read and also doesn't allow testing it.

Reply:
What do you think about something like this?

```
+    def load_fullchain(bundle_pem)
+      extra_chain_cert = File.open(bundle_pem)
+        .read.each_line
+        .slice_after(/END CERTIFICATE/)
+        .map(&:join)
+        .map { |pem| OpenSSL::X509::Certificate.new(pem) }
+      return extra_chain_cert
+    end
+
     def https_app(https_port, plugins = https_plugins)
       unless https_enabled?
         logger.warn "Missing SSL setup, https is disabled."
@@ -103,6 +112,7 @@ module Proxy
         :SSLPrivateKey => load_ssl_private_key(settings.ssl_private_key),
         :SSLCertificate => load_ssl_certificate(settings.ssl_certificate),
         :SSLCACertificateFile => settings.ssl_ca_file,
+       :SSLExtraChainCert => load_fullchain(settings.foreman_ssl_ca),
         :SSLOptions => ssl_options,
         :SSLCiphers => CIPHERS - Proxy::SETTINGS.ssl_disabled_ciphers,
         :daemonize => false,

```

Reply:
You don't need to return in Ruby and you can also use `File.read`. There's also no need to call map twice. For me this version works:
```ruby
def load_fullchain(bundle_pem)
  File.read(bundle_pem)
    .lines
    .slice_after(/END CERTIFICATE/)
    .map { |pem| OpenSSL::X509::Certificate.new(pem.join) }
end
```

Reply:
Ack, thanks for the guidance. I tested it here and it works fine. 

I will also add a verification that the bundle file (settings.foreman_ssl_ca) can be read, just like we already do for settings.ssl_ca_file


```
      unless File.readable?(settings.ssl_ca_file)
        logger.error "Unable to read #{settings.ssl_ca_file}. Are the values correct in settings.yml and do permissions allow reading?"
      end

      unless File.readable?(settings.foreman_ssl_ca)
        logger.error "Unable to read #{settings.foreman_ssl_ca}. Are the values correct in settings.yml and do permissions allow reading?"
      end
```